File: 240620/주사위 던지기/roll-a-dice.cpp
#include <iostream>
#include <tuple>
#include <utility>
using namespace std;
#define MAX_N 100
#define ASCII 128
#define DIR_NUM 4
#define OUT_OF_GRID make_pair(-1, -1)

int n, m, x, y;
int grid[MAX_N][MAX_N];
int dir_mapper[ASCII];

int u=1, f=2, r=3;

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<n;
}

pair<int, int> NextPos(int x, int y, int move_dir) {
    int dx[DIR_NUM] = {0, 0, -1, 1}, dy[DIR_NUM] = {1, -1, 0, 0};
    
    int nx = x+dx[move_dir], ny=y+dy[move_dir];
    if(InRange(nx, ny))
        return make_pair(nx, ny);
    else
        return OUT_OF_GRID;
}

void Simulate(int move_dir) {
    pair<int, int> next_pos = NextPos(x, y, move_dir);

    if(next_pos == OUT_OF_GRID) return;

    tie(x, y) = next_pos;

    if(move_dir == 0) {
        tie(u, f, r) = make_tuple(7-r, f, u);
    } else if (move_dir == 1) {
        tie(u, f, r) = make_tuple(r, f, 7-u);
    } else if (move_dir == 2) {
        tie(u, f, r) = make_tuple(f, 7-u, r);
    } else {
        tie(u, f, r) = make_tuple(7-f, u, r);
    }

    int down = 7-u;
    grid[x][y] = down;
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>m>>x>>y;
    x--; y--;

    dir_mapper['R'] = 0;
    dir_mapper['L'] = 1;
    dir_mapper['U'] = 2;
    dir_mapper['D'] = 3;

    grid[x][y] = 6;

    for(int i=0; i<m; i++) {
        char char_dir;
        cin >> char_dir;

        Simulate(dir_mapper[char_dir]);
    }

    int ans = 0;

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            ans += grid[i][j];
        }
    }

    cout << ans;

    return 0;
}

File: 250404/주사위 던지기/roll-a-dice.py
OUT_OF_GRID = (-1, -1)

n, m, x, y = tuple(map(int, input().split()))
grid = [[0 for _ in range(n)] for _ in range(n)]
movement = input().split()

up, front, right = 1, 2, 3


def next_pos(x, y, move_dir):
    dxs, dys = [0, 0, -1, 1], [1, -1, 0, 0]

    nx, ny = x+dxs[move_dir], y+dys[move_dir]
    return (nx, ny) if 0<=nx<n and 0<=ny<n else OUT_OF_GRID


def simulate(move_dir):
    global x, y, up, front, right

    nx, ny = next_pos(x, y, move_dir)

    if (nx, ny) == OUT_OF_GRID:
        return

    x, y = nx, ny
    
    if move_dir == 0:
        up, front, right = 7-right, front, up
    elif move_dir == 1:
        up, front, right = right, front, 7-up
    elif move_dir == 2:
        up, front, right = front, 7-up, right
    else:
        up, front, right = 7-front, up, right

    bottom = 7-up
    grid[x][y] = bottom


x -= 1
y -= 1

dir_mapper = {
    'R':0,
    'L':1,
    'U':2,
    'D':3
}

grid[x][y] = 6
for char_dir in movement:
    simulate(dir_mapper[char_dir])

ans = sum([
    grid[i][j]
    for i in range(n)
    for j in range(n)
])

print(ans)